src/ucs/sys/topo.h

```diff
@@ -0,0 +1,64 @@
+/**
+* Copyright (C) NVIDIA Corporation. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_TOPO_H
+#define UCS_TOPO_H
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include <ucs/type/status.h>
+#include <limits.h>
+
+BEGIN_C_DECLS
+
+/** @file topo.h */
+
+/**
+ * System device type
+ */
+typedef enum {
+    UCS_SYS_DEVICE_IB = 0, /* Infiniband Device */
+    UCS_SYS_DEVICE_CUDA,   /* NVIDIA GPU Device */
+    UCS_SYS_DEVICE_NUMA,   /* NUMA Memory */
+    UCS_SYS_DEVICE_LAST
+} ucs_sys_device_enum_t;
+
+
+/**
+ * @ingroup UCS_RESOURCE
+ * System Device abstraction
+ */
+typedef struct ucs_sys_device {
+    ucs_sys_device_enum_t sys_dev_type;             /**< Type of system device*/
+    unsigned int          id;                       /**< Index of the unit */
+    unsigned int          bus_id;                   /**< bus ID of of the device if applicable*/
+} ucs_sys_device_t;
+
+
+/**
+ * Detect system devices such as HCAs, GPUs, and other PCIe devices
+ * and return a list of such devices
+ *
+ * @param [out] sys_devices     Array of system device abstractions
+ * @param [out] num_sys_devices Number of system deivces detected in the system
+ * @return UCS_OK or error in case of failure.
+ */
+ucs_status_t ucs_topo_get_sys_devices(ucs_sys_device_t **sys_devices, int *num_sys_devices);
```

Comment:
why need to allocate/release sys devices?
maybe can just have a global read only structure?

Reply:
@yosefe So the suggestion is to populate the list of system devices (say during `ucp_init_version()`) and let that list be globally accessible? (maybe similar to `ucp_am_handlers`?) I assumed that global structures aren't favored in general but if that's not the case I can drop the alloc/release calls. 

In a multi-threaded application, threads need to explicitly synchronize on the success of ucp_init before any of them can call UCP API right?

Reply:
i think it can be global in UCS layer, created on 1st access. since the structure is expected to be read-only and reflect general system information, i think global is not a problem.